### PR TITLE
Layer refactorying

### PR DESCRIPTION
**Pull request by benson31**
Adds a bunch of layers to the factory. Once #1378 merges, I'll have to update this, I'm sure.

### CODE REVIEW

File: model_zoo/lbann.cpp [hunk 1]
```diff
@@ -132,10 +132,13 @@ int main(int argc, char *argv[]) {
     }
     El::ReportException(e);
     return EXIT_FAILURE;
-  } catch (std::exception& e) {
+  }
+#if 0
```
Comment (bvanessen) [inline model_zoo/lbann.cpp]: Is this to allow normal exceptions to propagate all of the way out?

Comment (benson31) [reply model_zoo/lbann.cpp]: That was an accidental commit, but yes. I had some `bad_cast` problems I was debugging. I'll remove this.

File: src/layers/learning/entrywise_scale_bias.cpp [hunk 2]
```diff
@@ -143,13 +143,17 @@ void entrywise_scale_bias_layer<TensorDataType, Layout, Device>::bp_compute() {
...
+#define PROTO(T)                                                        \
+  template class entrywise_scale_bias_layer<                            \
+    T, data_layout::DATA_PARALLEL, El::Device::CPU>;                    \
+  template class entrywise_scale_bias_layer<                            \
+    T, data_layout::MODEL_PARALLEL, El::Device::CPU>;                   \
+  LBANN_LAYER_BUILDER_ETI(entrywise_scale_bias, T, El::Device::CPU)
 
 #define LBANN_INSTANTIATE_CPU_HALF
 #include "lbann/macros/instantiate.hpp"
+#undef PROTO
```
Comment (bvanessen) [inline src/layers/learning/entrywise_scale_bias.cpp]: Do we need this undef here?

File: src/layers/transform/concatenate.cpp [hunk 3]
```diff
@@ -326,5 +326,6 @@ void bp_compute_impl(
 
 #define LBANN_INSTANTIATE_CPU_HALF
 #include "lbann/macros/instantiate.hpp"
+#undef PROTO
```
Comment (bvanessen) [inline src/layers/transform/concatenate.cpp]: I thought that we didn't need to do this in cpp files.

File: src/layers/transform/concatenate.cu [hunk 4]
```diff
@@ -470,13 +470,13 @@ void bp_compute_impl(
...
-#define PROTO(T)                                        \
-  template class concatenate_layer<                     \
-    T, data_layout::DATA_PARALLEL, El::Device::GPU>;    \
-  template class concatenate_layer<                     \
-    T, data_layout::MODEL_PARALLEL, El::Device::GPU>
+#define PROTO(T)                                                        \
+  template class concatenate_layer<                                     \
+    T, data_layout::DATA_PARALLEL, El::Device::GPU>;                    \
+  template class concatenate_layer<                                     \
+    T, data_layout::MODEL_PARALLEL, El::Device::GPU>;
```
Comment (bvanessen) [inline src/layers/transform/concatenate.cu]: pretty sure you don't need the last semi-colon

File: src/layers/learning/entrywise_scale_bias.cpp (hunk 2 above)
Comment (benson31) [reply src/layers/learning/entrywise_scale_bias.cpp]: No, will remove.

File: src/layers/transform/concatenate.cpp (hunk 3 above)
Comment (benson31) [reply src/layers/transform/concatenate.cpp]: We don't. Will remove. (Artifact of some temporary changes)

File: src/layers/transform/concatenate.cu (hunk 4 above)
Comment (benson31) [reply src/layers/transform/concatenate.cu]: Correct, also an artifact of other changes.

**Comment by benson31** [comment]
I'm working on the merge with develop. There are some issues, but they're small.

File: include/lbann/utils/cudnn.hpp [hunk 5]
```diff
@@ -84,6 +84,23 @@ class Layer;
...
 namespace cudnn {
 
+template <typename T>
+struct ScalingParameterT
+{
+  using type = T;
+};
+
+template <typename T>
+using ParamType = typename ScalingParameterT<T>::type;
```
Comment (timmoon10) [inline include/lbann/utils/cudnn.hpp]: We should document and link to the [cuDNN docs](https://docs.nvidia.com/deeplearning/sdk/cudnn-archived/cudnn_701/cudnn-user-guide/index.html#scaling-parameters).

**Comment by timmoon10** [comment]
As we've discussed, I'm not thrilled about the proliferation of `_builder.cpp` files. I'll update my newer, more experimental layers (#1397, #1403, #1404) to keep everything contained in the main source file.

**Pull request merged**